### ISSUE TITLE
ldap3 not working with Python 3.10

### ISSUE THREAD

**Issue by thmsklngr** [cannatag/ldap3#993]
Hi,
today I installed Python 3.10 on my machine, created a virtual environment and installed ldap3 2.9.1 in this environment. Whenever I try to connect a server, which was working with Python 3.8, I get an error:
```
LDAPSocketOpenError: ("('socket ssl wrapping error: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)',)",)
```
The trace:
```
~/.virtualenvs/dupcheck-S0klZjP4/lib/python3.10/site-packages/ldap3/core/connection.py in bind(self, read_server_info, controls)
    587                 self._bind_controls = None
    588                 if self.closed:  # try to open connection if closed
--> 589                     self.open(read_server_info=False)
    590                 if self.authentication == ANONYMOUS:
    591                     if log_enabled(PROTOCOL):

~/.virtualenvs/dupcheck-S0klZjP4/lib/python3.10/site-packages/ldap3/strategy/sync.py in open(self, reset_usage, read_server_info)
     55 
     56     def open(self, reset_usage=True, read_server_info=True):
---> 57         BaseStrategy.open(self, reset_usage, read_server_info)
     58         if read_server_info and not self.connection._deferred_open:
     59             try:

~/.virtualenvs/dupcheck-S0klZjP4/lib/python3.10/site-packages/ldap3/strategy/base.py in open(self, reset_usage, read_server_info)
    144                         if log_enabled(ERROR):
    145                             log(ERROR, '<%s> for <%s>', str(exception_history[0][0]) + ' ' + str((exception_history[0][1])), self.connection)
--> 146                         raise exception_history[0][0]
    147                     else:
    148                         if log_enabled(ERROR):
```
I switched to a Python 3.8 environment and the same code was working as expected. If you need further information, please let me know.

Regards, Thomas

**Comment by zorn96** [comment]
hi @mosesontheweb ! this actually doesn't seem to be an ldap3 issue at all.
python 3.10 changes the minimum openssl version that's linked to python
https://docs.python.org/3/whatsnew/3.10.html
```
Important deprecations, removals or restrictions:

PEP 644, Require OpenSSL 1.1.1 or newer
```

the error you're seeing about a handshake failure actually originates from the python ssl module. OpenSSL 1.1.1 has stronger minimum requirements on TLS versions and ciphers unless you specifically build it from source with different configuration options.

this is, of course, just the default. you can change it in how you install python. here's an example from my system of changes even across 2 python3.8 installations
```
azariazornberg@Azarias-MacBook-Pro ms_active_directory % python3.8 --version
Python 3.8.7
azariazornberg@Azarias-MacBook-Pro ms_active_directory % /usr/bin/python3 --version
Python 3.8.2
azariazornberg@Azarias-MacBook-Pro ms_active_directory % python3.8 -c 'import ssl; print(ssl.OPENSSL_VERSION)'     

OpenSSL 1.1.1k  25 Mar 2021
azariazornberg@Azarias-MacBook-Pro ms_active_directory % /usr/bin/python3 -c 'import ssl; print(ssl.OPENSSL_VERSION)'
LibreSSL 2.8.3
```

so basically the library didn't break - your server doesn't support newer TLS version/ciphers, and the new python will just refuse to talk to it over _any protocol_ by default. so to fix it you can either update your server or your python

**Comment by asteven** [comment]
Is someone else ends up here while searching for a solution to this when dealing with an old ldap server which cannot be changed here's a workaround:

```
import ldap3

# Copied from urllib3.util.ssl_.DEFAULT_CIPHERS on python 3.10.2
# to workaround old tls ciphers used by ldap server.
tls_ciphers = 'ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20:ECDH+AESGCM:DH+AESGCM:ECDH+AES:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!eNULL:!MD5:!DSS'

tls = ldap3.Tls(ciphers=tls_ciphers)
server_url = 'ldaps://your.ldap.server'
server = ldap3.Server(server_url, port=636, use_ssl=True, tls=tls)
conn = ldap3.Connection(server,
    auto_bind=True,
    client_strategy=ldap3.SYNC,
    user=...,
    password=...,
    authentication=ldap3.SIMPLE,
    check_names=True)
```

**Comment by sscherfke** [comment]
@asteven Thanks for the workaround <3
